src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -100,13 +100,13 @@ private PerAnalyzerState GetAnalyzerState(DiagnosticAnalyzer analyzer)
         {
             await EnsureAnalyzerActionCountsInitializedAsync(driver, cancellationToken).ConfigureAwait(false);
 
-            lock (_gate)
+            using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
             {
-                OnCompilationEventsGenerated_NoLock(compilationEvents, driver, cancellationToken);
+                await OnCompilationEventsGenerated_NoLockAsync(compilationEvents, driver, cancellationToken).ConfigureAwait(false);
             }
         }
 
-        private void OnCompilationEventsGenerated_NoLock(ImmutableArray<CompilationEvent> compilationEvents, AnalyzerDriver driver, CancellationToken cancellationToken)
+        private async Task OnCompilationEventsGenerated_NoLockAsync(ImmutableArray<CompilationEvent> compilationEvents, AnalyzerDriver driver, CancellationToken cancellationToken)
         {
             Debug.Assert(_lazyAnalyzerActionCountsMap != null);
 
```

Comment:
Computing the analyzer action counts needs us to invoke the Initialize method and compilation start actions, which can happen only after compilation started events have been generated and processed. So `_lazyAnalyzerActionCountsMap` needs to be lazily computed and any code path invoking `OnCompilationEventsGenerated_NoLockAsync` must have executed `EnsureAnalyzerActionCountsInitializedAsync`.